src/SSTIOGroup.cpp

```diff
@@ -0,0 +1,824 @@
+                }
+                m_control_available[field_control_name] = {
+                    .controls = controls,
+                    .domain = domain_type,
+                    .units = units,
+                    .agg_function = Agg::select_first,
+                    .description = description
+                };
+            }
+        }
+    }
+
+    void SSTIOGroup::add_mmio_signals(const std::string &raw_name,
+                                      int domain_type, uint32_t register_offset,
+                                      std::map<std::string, sst_signal_mmio_field_s> &fields)
+    {
+        int num_domain = m_topo.num_domain(domain_type);
+
+        for (const auto &ff : fields) {
+            auto field_name = ff.first;
+            auto field_desc = ff.second;
+            auto write_value = field_desc.write_value;
+            auto begin_bit = field_desc.begin_bit;
+            auto end_bit = field_desc.end_bit;
+            auto description = field_desc.description;
+            auto behavior = field_desc.behavior;
+            auto units = field_desc.units;
+            double multiplier = field_desc.multiplier;
+
+            char hex[32];
+            snprintf(hex, 32, "0x%05" PRIx32, register_offset);
+            std::string raw_signal_name = raw_name + "_" + hex + "#";
+
+            // add raw signal for every domain index
+            auto it = m_signal_available.find(raw_signal_name);
+            if (it == m_signal_available.end()) {
+                std::vector<std::shared_ptr<Signal> > signals;
+                for (int domain_idx = 0; domain_idx < num_domain; ++domain_idx) {
+                    auto cpus = m_topo.domain_nested(GEOPM_DOMAIN_CPU, domain_type, domain_idx);
+                    int cpu_idx = *(cpus.begin());
+                    uint32_t augmented_offset = domain_type == GEOPM_DOMAIN_CORE
+                        ? register_offset + m_sstio->get_punit_from_cpu(domain_idx) * 4
+                        : register_offset;
+                    auto raw_sst = std::make_shared<SSTSignal>(
+                        m_sstio, SSTSignal::M_MMIO, cpu_idx, 0x00, 0x00, augmented_offset, write_value);
+
+                    signals.push_back(raw_sst);
+                }
+                m_signal_available[raw_signal_name] = {
+                    .signals = signals,
+                    .domain = domain_type,
+                    .units = units,
+                    .agg_function = Agg::select_first,
+                    .description = description,
+                    .behavior = behavior
+                };
+            }
+
+            std::string field_signal_name = raw_name + ":" + field_name;
+            auto raw_sst_list = m_signal_available.at(raw_signal_name).signals;
+            std::vector<std::shared_ptr<Signal> > signals;
+            for (int domain_idx = 0; domain_idx < num_domain; ++domain_idx) {
+                auto field_signal = std::make_shared<MSRFieldSignal>(
```

Comment:
Will this signal be visible to users with a name  prefixed by "MSR::", but provided by the "SST" IOGroup?  Would be nice if we could avoid that.

Reply:
This won't include MSR in the name. This is just reusing the code that a bitmasked field down to multiple signals. I'll add a comment to indicate that in the code.